### ISSUE TITLE
security contact / mutation xss

### ISSUE THREAD

**Issue by gmatuz** [bigskysoftware/htmx#1685]
what is your preferred way to be contacted for security issues?

**Comment by Telroshan** [comment]
Can't speak for @1cg there but I'd say GitHub issues are fine, so is the htmx Discord.

If that's something that really compromises every existing project using htmx, then probably even send a DM?
Otherwise, you might want to update that very GitHub issue we're talking in, renaming it & providing details about the security issues you've encountered!

**Comment by gmatuz** [comment]
Thanks @Telroshan the issue is not a biggie so I'll just keep it here!

I'm just getting into htmx, the assumption I'm making here is htmx considers escaping/sanitization part of _server side responsibilities_ expecting safe code to use something like [dompurify](https://github.com/cure53/DOMPurify) to sanitize or have templating (like JSX) take care of escaping. However, it should still be an issue if there are inputs that are normally benign or neutralised but they still trigger XSS when rendered by htmx 

## Mutation XSS
The way `<head>` is matched in the head support extension  creates options for mutation XSS (having input that would be understood as benign by most sanitisers passed to htmx which _mutates_ it to html with potentially attacker controlled javascript).
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/ext/head-support.js#L21
For example full html tags or `<>` are often considered benign in [attribute context ](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#output-encoding-for-html-attribute-contexts) but for example if there is user controlled data that gets into an attribute the attacker could set the value to something like `<head><script>alert(1337)</script></head>` which gets to the response as for example 
`<p title='<head><script>alert(1337)</script></head>'>xss</p>`
this usually would not be an issue (e.g. [dompurify](https://cure53.de/purify) will just leave it alone) but the matching logic will pick up the attribute value as a tag and include it in the header triggering the injection.
_If it was not for HTMX's mutation_ this would not normally result in an XSS (and I reckon it is also not an intentionally included new sink like hx-vars). I can provide a PoC if it is not clear; tested and it works.

Similarly removing  svg tags based on string match without the context (and knowing if the response is even valid html) can probably lead to mutation XSS. 
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/ext/head-support.js#L19
This replace can be used to interlace a split XSS attack vector or in some cases even possibly bypass encoding (have 2 parts of the vector bound in different places of a template and use the svg replace to remove anything in between).  I can provide a specific example for this too if necessary.

## General pattern and considerations
To avoid these types of confusions between the matching logic and the inclusion it is probably a good idea to use functions for matching that implicitly use the same parser logic instead of doing it explicitly. e.g. looking for a `<head>` with a query selector or comparing with the keys on the headers. (see below)..  

I'd argue this probably makes sense even from the perspective of robustness avoiding strange edge cases, let alone security.

There are a few more cases of potential confusion between how the browser will parse things vs the way it is matched in the code.

## Other examples
1. When checking for headers
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/htmx.js#L2808-L2810
given that none of the regexes used to invoke the function are anchored, e.g. 
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/htmx.js#L3186
the header string can be in the value of any of the headers (`something-else:  HX-Push:`)
This does not result in anything exploitable _currently_ since you do get the values as header values e.g.
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/htmx.js#L3186

2. Possible in the title replacement as well, but did not dig into this one yet
https://github.com/bigskysoftware/htmx/blob/47af93bdb0c9f5d356e36fd6073e0b74d12b9242/src/htmx.js#L1081-L1084

**Comment by gmatuz** [comment]
> Thank you @gmatuz for looking into this, an I see how the header extension could get into trouble, although typically wouldn't user content be escaped to `&lt;` rather than passed through raw as an attribute?

So `<` is not a metacharacter in the context of an _attribute_ so if you build some lib that does context aware escaping on sanitization and you want to avoid being opinionated and unnecessarily restrictive (which I guess would be a design goal in these cases) you would make sure to escape different quotes but leave other characters as is. That said I was playing around with JSX and that would escape `<` as well. So in short some in reality do but probably not all and in theory there is no need to. I'd say in your case there is an argument for expecting the most liberal but legitimately correct escaping/sanitization and work nicely with that.

> Do you have a fix you'd propose for head-support.js? We are considering doing something similar for htmx 2. Maybe just move over to a full document parse and extract the head element (we could still gate on a regex for perf reasons)

I cannot talk to the performance (I'll try to do some tests) but my first approach would be what you mentioned. Keep the match as a gate and parse the whole response. Alternatively you could make a more sophisticated regex that would somehow keep track of the context of the match.

**Comment by gmatuz** [comment]
> looking at the other instances, I do not think there can be script injection. If a hostile entity has control over unescaped header content then you are already in a pretty bad spot. The title is set directly via window.title, so there isn't an opportunity to inject anything, as far as I can tell.

I agree with the title part, document.title it is not a javascript sink so maybe somebody could change the title but not directly an XSS. I think there was another case that I did not dig into yet to see what sink you can potentially get to but maybe you can tell me about this here:
https://github.com/bigskysoftware/htmx/blob/9ed1275ec353c0c783044799e9393c5390d66f03/src/htmx.js#L1092-L1093C13

On the header part: being able to write to the value of a header is not necessarily a huge problem by itself (it really depends on what header and how it is used) and it does happen (not even very rare) that user supplied data gets there. It is slightly worse to get something like [CRLF injection](https://owasp.org/www-community/vulnerabilities/CRLF_Injection) when you can add a new line and start adding an arbitrary header. This case _can_ lead to XSS but usually you need something on the client side to work with you (like in your case having hx-vars evaled later). 
So it definitely does make an attack more powerful if you can go from writing a header value to an XSS because that is not something you can generally do. I think this is something you should keep in mind when working on frontend but it is not a very important escalation/trust boundary so not a biggie **and** in your case since you are reading the values correctly (based on the right key) somebody would have to still make another mistake. 

However in this case you could still consider making sure `hasHeader` only returns `true` if there is in fact a header like that: throwing in some logic that makes sure that the regex in anchored and match on lines or wraps the regex into forcing that it is either after crlf or the first char or checking if `getResponseHeader` is `null` but I reckon there is probably some reason I'm missing why you opted for this option instead in the first place.

**Comment by gmatuz** [comment]
actually you are correct @1cg , all template engines escape `<`

**Comment by gmatuz** [comment]
More interestingly though the head search trigger for JSON responses as well. I guess that is a more likely attack vector

**Comment by andryyy** [comment]
I can hardly follow this. It's like shooting yourself in the foot. It's not like someone who has access to the server response can to pretty much anything.

**Comment by gmatuz** [comment]
@andryyy  the assumption I'm trying to go on here is that the server correctly escapes things and trying to look at cases when it still results in XSS on the client side.

What I was explaining above is let's say I set the _title_ to `<head><script>alert(1337)</script></head>` in you [example](https://jsbin.com/rukosaxebo/edit?html,console,output) to  receive the following JSON as response `{ ... "title" : "<head><script>alert(1337)</script></head>"}` which _should be_ perfectly fine and expected because I will use client side templates to escape it eventually and it is not HTML anyway. 
However, the parsing logic will still trigger on it currently and swap in the head to trigger the javascript.
